core/Pos/Binary/Class/TH.hs

```diff
@@ -0,0 +1,460 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TemplateHaskell     #-}
+
+-- | TH helpers for Bi
+module Pos.Binary.Class.TH
+       ( deriveSimpleBi
+       , Cons (..)
+       , Field (Field)
+       ) where
+
+import           Universum
+import           Unsafe                   (unsafeHead)
+
+import           Control.Lens             (imap)
+import           Data.Default             (def)
+import           Data.List                (notElem, nubBy, partition)
+import qualified Data.Store.Internal      as Store
+import qualified Data.Text                as T
+import           Formatting               (sformat, shown, (%))
+import           Language.Haskell.TH
+import           TH.ReifySimple           (DataCon (..), DataType (..), reifyDataType)
+import           TH.Utilities             (plainInstanceD)
+
+import qualified Pos.Binary.Class.Core    as Bi
+
+{-
+Suppose you have the following datatype:
+
+data User
+    = Login {
+      login :: String
+    , age   :: Int
+    }
+    | FullName {
+      firstName  :: String
+    , lastName   :: String
+    , sex        :: Bool
+    }
+
+then the next deriveSimpleBi:
+
+deriveSimpleBi ''User [
+    Cons 'Login [
+        Field 'login ''String,
```

Comment:
we were discussing it w/ @neongreen and decided it's more clear for serialization: you can look at this instance without the necessity to look at datatype definition